### PR TITLE
Adding an e2e test for admission webhook

### PR DESCRIPTION
**Pull request by caesarxuchao**
Part of https://github.com/kubernetes/features/issues/492

The purpose of this test is making sure the webhooks get called, and the apiserver can communicate with the webhook.

We will expand the test cover more webhook features in followups.

The webhook used in the test rejects pods with container names "webhook-disallow". Will upload the source code of the example in a follow up PR.

### CODE REVIEW

File: cluster/gce/config-default.sh
Comment (caesarxuchao) [inline cluster/gce/config-default.sh]: This doesn't seem to be useful, e2e test says
```
I1018 22:29:18.099] [It] Should be able to deny pod creation
I1018 22:29:18.099]   test/e2e/apimachinery/webhook.go:56
I1018 22:29:18.100] Oct 18 22:28:53.647: INFO: dynamic configuration of initializers requires the alpha admissionregistration.k8s.io group to be enabled
```

Comment (krzyzacy) [reply cluster/gce/config-default.sh]: change it to `RUNTIME_CONFIG=admissionregistration.k8s.io/v1alpha1`, or else it will still read in `KUBE_RUNTIME_CONFIG`

File: cluster/gce/config-default.sh
Comment (caesarxuchao) [inline cluster/gce/config-default.sh]: Got the same result. In apiserver.log:
```
I1018 23:08:04.489359       5 master.go:407] Skipping disabled API group "admissionregistration.k8s.io".
````
@krzyzacy help?

**Comment by caesarxuchao** [comment]
Webhook is broken at HEAD...

**Comment by caesarxuchao** [comment]
cc @deads2k this is the basic webhook e2e test that I want to get in quickly.

File: test/e2e/apimachinery/webhook.go
Comment (cheftako) [inline test/e2e/apimachinery/webhook.go]: We are trying to remove explicit creation of namespace to run test in. (Please see https://github.com/kubernetes/kubernetes/pull/53479) Its causing problem with some test framework improvements which are in the works and I don't believe is really needed.

**Comment by deads2k** [comment]
> cc @deads2k this is the basic webhook e2e test that I want to get in quickly.

While I agree that it's a thing that we want, I think that wiring it now against a configuration model that has known problems and is going to change dramatically is going to be a problem.  We need address the topology concerns before we can reasonably wire an e2e.

**Comment by lavalamp** [comment]
We need to keep the codebase in a working state. I think it should be OK to merge this. It shouldn't be considered "extra" work to keep the tests functioning when we make changes.

File: test/e2e/apimachinery/webhook.go
Comment (lavalamp) [reply test/e2e/apimachinery/webhook.go]: Shouldn't the framework be creating this for you?

File: test/e2e/apimachinery/webhook.go
Comment (lavalamp) [inline test/e2e/apimachinery/webhook.go]: does this need to be fixed before this can merge?

**Comment by lavalamp** [comment]
This LGTM when the comments are addressed.

**Comment by deads2k** [comment]
/hold

This is pinning an authentication flag and cert binding structure which is likely to change as a result of https://github.com/kubernetes/kubernetes/issues/54404 and pinning that structure now doesn't serve us well.  I'm in favor of tests, but not tests which are deliberately pinning an authentication story that has known issues and known solutions already open.

Once https://github.com/kubernetes/kubernetes/issues/54404 is settled, making a test that uses the authentication story is a great addition.

File: test/e2e/apimachinery/webhook.go
Comment (caesarxuchao) [reply test/e2e/apimachinery/webhook.go]: Done.

File: test/e2e/apimachinery/webhook.go
Comment (caesarxuchao) [reply test/e2e/apimachinery/webhook.go]: No. We'll expand the test cases when the apiserver sends external objects.

**Comment by caesarxuchao** [comment]
Checked in the webhook source code at /test/image, in commit "webhook source code".

File: test/e2e/apimachinery/webhook.go
Comment (cheftako) [inline test/e2e/apimachinery/webhook.go]: Might be worth documenting if this is a backward compatibility test. i.e. do we want to maintain a build using 1.8 to ensure that old webhooks work with new k8s clusters?

Comment (liggitt) [reply test/e2e/apimachinery/webhook.go]: a webhook that works with 1.8 would have to be expecting internal objects, right? that's not something we should maintain compatibility with

**Comment by deads2k** [comment]
/unhold

now that https://github.com/kubernetes/kubernetes/pull/54414 merged.

**Comment by caesarxuchao** [comment]
I removed the mutual tls requirement from the webhook image. The test will still verify other parts of the webhook feature.

I'll open an issue reminding ourselves to add the authn back after we reached agreement on what CA to use (#53828) and figure out how to plumb the CA to the `--admission-config-file` during e2e test.

File: test/e2e/apimachinery/webhook.go
Comment (cheftako) [inline test/e2e/apimachinery/webhook.go]: It seems like this highlights the problem that we have no API to tell us which webhooks are enabled and functioning. I believe we are generally going to want that functionality and once we have it we can fix this sleep.

**Review by cheftako — APPROVED** [review]

**Comment by caesarxuchao** [comment]
Tracking the client authentication in https://github.com/kubernetes/kubernetes/issues/54709.

File: test/e2e/apimachinery/webhook.go
Comment (caesarxuchao) [reply test/e2e/apimachinery/webhook.go]: Tracked in https://github.com/kubernetes/kubernetes/issues/54712

**Comment by caesarxuchao** [comment]
@deads2k @lavalamp could you approve? Thank you.

**Comment by caesarxuchao** [comment]
Rebased.

**Pull request merged**
